engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,841 @@
+		c.tables[targetTable] = make(map[metadata.SourceTable]struct{}, 0)
+		for _, sourceTable := range sourceTables {
+			c.tables[targetTable][sourceTable] = struct{}{}
+		}
+	}
+	return nil
+}
+
+// Coordinate coordinates ddls.
+func (c *DDLCoordinator) Coordinate(ctx context.Context, item *metadata.DDLItem) ([]string, optimism.ConflictStage, error) {
+	c.mu.Lock()
+	c.pendings.Add(1)
+	defer c.pendings.Done()
+	c.mu.Unlock()
+
+	jobCfg, err := c.jobStore.GetJobCfg(ctx)
+	if err != nil {
+		return nil, optimism.ConflictError, err
+	} else if jobCfg.ShardMode == "" {
+		return nil, optimism.ConflictError, errors.New("coordinate error with non-shard-mode")
+	}
+
+	// create shard group if not exists.
+	g, err := c.loadOrCreateShardGroup(ctx, item.TargetTable, jobCfg)
+	if err != nil {
+		return nil, optimism.ConflictError, err
+	}
+
+	ddls, conflictStage, deleted, err := g.handle(ctx, item)
+	if err != nil {
+		return ddls, conflictStage, err
+	}
+
+	// if all source table is deleted, we should remove the shard group.
+	if deleted {
+		c.removeShardGroup(ctx, item.TargetTable)
+	}
+
+	// handle table level ddl
+	switch item.Type {
+	case metadata.CreateTable:
+		c.mu.Lock()
+		tables, ok := c.tables[item.TargetTable]
+		if !ok {
+			c.tables[item.TargetTable] = make(map[metadata.SourceTable]struct{}, 0)
+			tables = c.tables[item.TargetTable]
+		}
+		tables[item.SourceTable] = struct{}{}
+		c.mu.Unlock()
+	case metadata.DropTable:
+		c.mu.Lock()
+		delete(c.tables[item.TargetTable], item.SourceTable)
+		if len(c.tables[item.TargetTable]) == 0 {
+			delete(c.tables, item.TargetTable)
+		}
+		c.mu.Unlock()
+	}
+	return ddls, conflictStage, err
+}
+
+// ShowDDLLocks show ddl locks.
+func (c *DDLCoordinator) ShowDDLLocks(ctx context.Context) ShowDDLLocksResponse {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	ddlLocks := make(map[metadata.TargetTable]DDLLock)
+	for targetTable, g := range c.shardGroups {
+		tbs := g.showTables()
+		ddlLocks[targetTable] = DDLLock{ShardTables: tbs}
+	}
+	return ShowDDLLocksResponse{Locks: ddlLocks}
+}
+
+func (c *DDLCoordinator) loadOrCreateShardGroup(ctx context.Context, targetTable metadata.TargetTable, jobCfg *config.JobCfg) (*shardGroup, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if g, ok := c.shardGroups[targetTable]; ok {
+		return g, nil
+	}
+
+	newGroup, err := newShardGroup(ctx, c.jobID, jobCfg, targetTable, c.tables[targetTable], c.kvClient, c.tableAgent)
+	if err != nil {
+		return nil, err
+	}
+	c.shardGroups[targetTable] = newGroup
+	return newGroup, nil
+}
+
+func (c *DDLCoordinator) removeShardGroup(ctx context.Context, targetTable metadata.TargetTable) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if g, ok := c.shardGroups[targetTable]; ok {
+		if err := g.clear(ctx); err != nil {
+			c.logger.Error("clear shard group failed", zap.Error(err))
+		}
+	}
+	delete(c.shardGroups, targetTable)
+}
+
+type shardGroup struct {
+	mu                  sync.RWMutex
+	normalTables        map[metadata.SourceTable]string
```

Comment:
please left some comment about the meaning of `string` value and what's a normal/conflict table